src/music/Scales.ts

```diff
@@ -0,0 +1,140 @@
+            if (curr & 1) { // only add proper scales - i.e., ones with roots
+                r.push(curr);
+            }
+            curr = rotateScale(curr, 1); // advance to the next mode
+        }
+        return r;
+    }
+
+    export function hasInterval(scale: Scale, semitones: number): boolean {
+        semitones = mod(semitones, 12);
+        return (scale & (1 << semitones)) > 0; // if the semitones-th bit is set, the scale contains a given interval
+    }
+
+    // count the instances of a particular interval in a scale
+    // useful for counting tritonia / hemitonia / etc.
+    export function countInterval(scale: Scale, semitones: number): number {
+        semitones = mod(semitones, 12);
+        const count: number = getModes(scale) // starting from each note in the scale...
+            .filter(mode => hasInterval(mode, semitones)).length; // count all the notes with a note N semitones above them
+        if (semitones === 6) { // special case (tritones get counted twice since they're symmetrical)
+            return count / 2;
+        }
+        return count;
+    }
+
+    // count how many notes in the scale lack a perfect fifth above them
+    export function getImperfections(scale: Scale): number {
+        return getModes(scale).filter(mode => !hasInterval(mode, 7)).length;
+    }
+
+    export function containsScale(parent: Scale, child: Scale): boolean {
+        return (parent & child) === child;
+    }
+
+    export function containsChord(scale: Scale, chord: number[]): boolean {
+        return containsScale(scale, createScaleFromPitchClass(chord));
+    }
+
+    const scales: Scale[] = Arrays.flatten([
+        getModes(0xab5), // major
+        getModes(0x9ab), // neapolitan minor
+        getModes(0x557), // leading whole-tone inverse
+        getModes(0x9ad), // harmonic minor
+        getModes(0x9b5), // harmonic major
+        getModes(0xaad), // melodic minor
+        getModes(0xa6d), // Jeths' mode
+        getModes(0x9b3), // double harmonic major
+    ]);
+
+    // TODO: this can be made more efficient by caching modes.
+    // see if that's worthwhile.
+    export function matchesQuery(scale: Scale, query: ScaleQuery): boolean {
+        if (query.notes) {
+            const notes = getNumberOfNotes(scale);
+            if (notes < query.notes[0] || notes > query.notes[1]) {
+                return false;
+            }
+        }
+        if (query.imperfections) {
+            const imperfections = getImperfections(scale);
+            if (imperfections < query.imperfections[0] || imperfections > query.imperfections[1]) {
+                return false;
+            }
+        }
+        if (query.hemitones) {
+            const hemitones = countInterval(scale, 1);
+            if (hemitones < query.hemitones[0] || hemitones > query.hemitones[1]) {
+                return false;
+            }
+        }
+        if (query.chord && !containsScale(scale, query.chord)) {
```

Comment:
`query.chord` is of type `number | undefined`, so this truthiness check scares me

Reply:
I'm not super worried about the case where `query.chord` is zero. Since scales are bit vectors, one scale `a` containing another scale `b` is expressed as `(a & b) = b`. This is trivially true for `b = 0`, so every scale passes the query (which is equivalent to just not checking the chord at all.)